test/test_dataloader.py

```diff
@@ -3673,7 +3724,1264 @@ def test_out_of_order_iterable_ds(self):
+        self.size = size
+        self.data = [{"id": i, "value": i * 2} for i in range(size)]
+        self.shuffle = shuffle
+
+    def __len__(self):
+        return self.size
+
+    def __getitem__(self, i):
+        if self.shuffle:
+            i = torch.randint(self.size, (1,)).item()
+        return self.data[i]
+
+    def state_dict(self):
+        if self.shuffle:
+            return {"rng_state": torch.get_rng_state()}
+        else:
+            return {}
+
+    def load_state_dict(self, state_dict):
+        if self.shuffle and "rng_state" in state_dict:
+            torch.set_rng_state(state_dict["rng_state"])
+
+
+@unittest.skipIf(
+    TEST_WITH_TSAN,
+    "Fails with TSAN with the following error: starting new threads after multi-threaded "
+    "fork is not supported. Dying (set die_after_fork=0 to override)",
+)
+class TestStatefulDataLoaderMapDataset(TestCase):
+    """Test stateful DataLoader with map-style datasets."""
+
+    @parametrize("num_workers", [0])
+    def test_map_dataset_statefulness(self, num_workers):
+        """Test state_dict/load_state_dict functionality with map datasets"""
+
+        dataset = StatefulMapDataset(50, shuffle=False)
+        dl = DataLoader(
+            dataset=dataset,
+            num_workers=num_workers,
+            collate_fn=identity_collate,
+            stateful=True,
+            batch_size=5,
+        )
+
+        # Consume some batches and track expected results
+        batches_before_checkpoint = []
+        it = iter(dl)
+        for i in range(3):  # consume first 3 batches
+            batch = next(it)
+            batches_before_checkpoint.append(batch)
+
+        # Save state after consuming 3 batches
+        state_dict = dl.state_dict()
+        self.assertIsInstance(state_dict, dict)
+
+        # Continue with original iterator and collect remaining batches
+        remaining_batches_original = []
+        for batch in it:
+            remaining_batches_original.append(batch)
+
+        # Create new loader and resume from checkpoint
+        dataset2 = StatefulMapDataset(50, shuffle=False)  # same dataset
+        dl2 = DataLoader(
+            dataset=dataset2,
+            num_workers=num_workers,
+            collate_fn=identity_collate,
+            stateful=True,
+            batch_size=5,
+        )
+        dl2.load_state_dict(state_dict)
+
+        # Collect all batches from resumed loader
+        remaining_batches_resumed = []
+        for batch in dl2:
+            remaining_batches_resumed.append(batch)
+
+        # Verify that resumed loader continues exactly where original left off
+        self.assertEqual(
+            len(remaining_batches_original), len(remaining_batches_resumed)
+        )
+        for orig, resumed in zip(remaining_batches_original, remaining_batches_resumed):
+            self.assertEqual(len(orig), len(resumed))
+            for o, r in zip(orig, resumed):
+                self.assertEqual(o, r)
```

Comment:
Since you're using `identity_collate` (i.e. you don't have tensors), should be able to just assertEqual the whole thing
```suggestion
        self.assertEqual(remaining_batches_original, remaining_batches_resumed)
```